client/asset/eth/eth.go

```diff
@@ -204,11 +204,102 @@ type ethFetcher interface {
 	sendToAddr(ctx context.Context, addr common.Address, val uint64) (*types.Transaction, error)
 	transactionConfirmations(context.Context, common.Hash) (uint32, error)
 	sendSignedTransaction(ctx context.Context, tx *types.Transaction) error
+	netFeeState(ctx context.Context) (baseFees, tipCap *big.Int, err error)
 }
 
 // Check that ExchangeWallet satisfies the asset.Wallet interface.
 var _ asset.Wallet = (*ExchangeWallet)(nil)
-var _ asset.AccountRedeemer = (*ExchangeWallet)(nil)
+var _ asset.AccountLocker = (*ExchangeWallet)(nil)
+var _ asset.FeeRater = (*ExchangeWallet)(nil)
+
+// fundReserveType represents the various uses for which funds need to be locked:
+// initiations, redemptions, and refunds.
+type fundReserveType uint32
+
+const (
+	initiationReserve fundReserveType = iota
+	redemptionReserve
+	refundReserve
+)
+
+func (f fundReserveType) String() string {
+	switch f {
+	case initiationReserve:
+		return "initiation"
+	case redemptionReserve:
+		return "redemption"
+	case refundReserve:
+		return "refund"
+	default:
+		return ""
+	}
+}
+
+// fundReserves is a struct which keeps track of the funds locked for
+// initiations, redemptions, and refunds.
+type fundReserves struct {
+	initiateReserves   uint64
+	redemptionReserves uint64
+	refundReserves     uint64
+}
+
+// fundReserveOfType returns a pointer to the funds reserved for a particular
+// use case.
+func (r *fundReserves) fundReserveOfType(t fundReserveType) (*uint64, error) {
+	switch t {
+	case initiationReserve:
+		return &r.initiateReserves, nil
+	case redemptionReserve:
+		return &r.redemptionReserves, nil
+	case refundReserve:
+		return &r.refundReserves, nil
+	default:
+		return nil, fmt.Errorf("invalid fund reserve type: %v", t)
```

Comment:
I'm tempted to say panic, remove the error and logging in `unlock`'s caller.
With an enum of `fundReserveType`s, it takes a special type of programmer error to pass an invalid value.